src/widgets/WidgetApi.ts

```diff
@@ -98,6 +108,16 @@ export class WidgetApi {
 
                     // Automatically acknowledge so we can move on
                     this.replyToRequest(<ToWidgetRequest>payload, {});
+                } else if (payload.action === KnownWidgetActions.Terminate) {
+                    // Finalization needs to be async, so postpone with a promise
+                    let finalizePromise = Promise.resolve();
+                    const wait = promise => {
+                        finalizePromise = finalizePromise.then(value => promise);
+                    }
+                    this.emit('terminate', wait);
+                    Promise.resolve(finalizePromise).then(() => {
+                        this.replyToRequest(<ToWidgetRequest>payload, {});
```

Comment:
We should probably comment what we're doing here:
```suggestion
                        // Acknowledge that we're shut down now
                        this.replyToRequest(<ToWidgetRequest>payload, {});
```